### ISSUE TITLE
Computation of Impulse Response (IR) in biophysics-based linearized os

### ISSUE THREAD

**Issue by npcottaris** [isetbio/isetbio#195]
In the linearized os model, the `osLinear.linearFilters()` method computes IR functions using an isomerization stimulus with the target background level on which a pulse of 1 photon / os.timeStep is superimposed. For the default 0.1ms time step, this corresponds to an impulse size of 10,000 R*/sec.

I am raising a few points for discussion.
- Is this an appropriate level for computing the IR, or should we be using a smaller stimulus? In traditional system characterization, one wants to use a level at which the system is mostly linear.

- In the current implementation, the derived IRs depend on os.timeStep, as the stimulus specifies pulse consisting of a 1 photon/time step. Do we want this scheme, or do we want to use a step specified as a constant isomerization rate, which is independent of the timeStep?

- Also, should we use the same pulse amplitude to derive the S-cone IR, despite the fact that S-cone mean activation is about 8 times lower than L/M cones? An other approach is to use a constant isomerization rate pulse, say 10% above the background. The stimuli used to derived the L-, M-, and S-cone IRs for one particular background level are shown in the figure below. 

![pulses](https://cloud.githubusercontent.com/assets/3444212/20932236/f551b928-bba1-11e6-9304-32e21d382035.png)

**Comment by fmrieke** [comment]
These are good points.  Although the implied rate is very high for the test stimulus, it is maintained for such a short period that it should be in the linear range (i.e. it is one isomerization total if I understand how it works properly).  Using a single isomerization should also make it independent of the time step in the calculation as long as the time step is short compared to the response (which it has to be for the biophysical model to work at all).  So I think it is fine to have it as 1 isomerization per time step given that the time step is short.  It would be good to verify linearity - i.e. to try a test stimulus twice as large and make sure we get a response scaled 2x but otherwise unchanged.  

Fred

**Comment by npcottaris** [comment]
Thanks Fred.

I verified that for a background luminance of 200 cd/m2, a pulse amplitude specified as 1 photon/time bin and a pulse amplitude specified as a 20% modulation of the background give essentially identical IRs. You start getting different IR shapes with pulse stimuli that are much stronger, like 100 photons/time bin. So we can out this issue to rest. I did  find and fixed a bug in linearFilters though.

**Comment by npcottaris** [comment]
We need to set a new default integration time for the coneMosaic object. 
Based on the following graph which shows the downsampled linear filters for 5, 15 and 15 ms (R,G,B respectively) and the original IR (for os.timeStep = 0.1ms) in gray, I suggest we set the default value to less than 25 ms. I propose 10ms. Let's make a decision on this asap.

![integrationtime](https://cloud.githubusercontent.com/assets/3444212/21014670/efe01642-bd2c-11e6-9766-3140d25d30d1.png)

**Comment by wandell** [comment]
Agreed.  I think I did make this change in the bipolar branch, but maybe I am just hallucinating.
10ms is fine.  I think I used 5ms.  

I kind of remember being surprised and pleased that the validations still passed.

**Comment by npcottaris** [comment]
Great. I will wait then until we merge the bipolar branch to master.

We should also throw warnings when the user:
* sets the `coneMosaic.integrationTime` > 25ms 
* sets the `os.timeStep` > 1 ms

We should set those limits as constants in `@coneMosaic` and in `@outerSegment`, respectively, and throw errors in the corresponding setters. This is how I handled the `noiseFlag` in `@coneMosaic`and in @outerSegment:

~~~
properties (Constant)
   validNoiseFlags = {'random','none','frozen'};
end

function set.noiseFlag(obj, val)
  if ischar(val) && (ismember(lower(val), coneMosaic.validNoiseFlags))
     obj.noiseFlag = val;
  else
     validNoiseFlags = coneMosaic.validNoiseFlags
     illegalValueForNoiseFlag = val
     error('This is an invalid value for coneMosaic.noiseFlag');
  end
end
~~~

This also allows the user to determine valid values for the `noiseFlag` without looking at the code. 
All he/she has to do is type at Matlab's command window:
~~~
coneMosaic.validNoiseFlags
~~~
which results in:
~~~
ans = 

    'random'    'none'    'frozen'
~~~

**Comment by wandell** [comment]
Thanks.  Small things.

- I did another merge of master into bipolar, which I do from time to time.  There were some conflicts.  I think I resolved the code (including validation scripts).  They all run in bipolar, though the v_os* do not match the validation data (as we know, already).
- I like the definition for validNoiseFlag() as a string. 
- I guess I am OK with warning on those parameters.  Mixed feelings, but you are probably right.

I would like to merge bipolar into master because all there are so many changes that the branch is ridiculously out of sync with master.  I will do a pull request.  Then we can start new (and hopefully much smaller) branches.  Though there will be one more big RGC push.
